tket/src/Transformations/SingleQubitSquash.cpp

```diff
@@ -0,0 +1,254 @@
+        sub = pair.first;
+        Gate_ptr left_over_gate = pair.second;
+        if (left_over_gate != nullptr) {
+          // => commute leftover through before squashing
+          insert_left_over_gate(left_over_gate, next_edge(v, e), condition);
+          left_over_gate = nullptr;
+        }
+        if (reversed_) {
+          sub = sub.dagger();
+        }
+
+        // we squash if the replacement is at least as good as the original
+        // (and it's not a no-op)
+        if (sub_is_better(sub, single_chain)) {
+          substitute(sub, bin, e, condition);
+          success = true;
+        }
+      }
+    }
+    if (e == out || is_last_optype(v_type)) {
+      squasher_->clear();
+      break;
+    }
+    if (move_to_next_vertex) {
+      if (is_gate_type(v_type)) {
+        bin.push_back(v);
+        single_chain.push_back(as_gate_ptr(v_op));
+      }
+      e = next_edge(v, e);
+      v = next_vertex(e);
+    }
+    if (reset_search) {
+      bin.clear();
+      single_chain.clear();
+      squasher_->clear();
+      condition = std::nullopt;
+    }
+  }
+  return success;
+}
+
+void SingleQubitSquash::substitute(
+    const Circuit &sub, const VertexVec &single_chain, Edge &e,
+    const Condition &condition) {
+  // backup edge
+  VertPort backup = {next_vertex(e), next_port(e)};
+
+  if (condition) {
+    circ_ptr_->substitute_conditional(
+        sub, single_chain.front(), Circuit::VertexDeletion::No);
+  } else {
+    circ_ptr_->substitute(
+        sub, single_chain.front(), Circuit::VertexDeletion::No);
+  }
+  circ_ptr_->remove_vertices(
+      VertexSet{single_chain.begin(), single_chain.end()},
+      Circuit::GraphRewiring::Yes, Circuit::VertexDeletion::Yes);
+
+  // restore backup
+  e = prev_edge(backup);
+}
+
+void SingleQubitSquash::insert_left_over_gate(
+    Op_ptr left_over, const Edge &e, const Condition &condition) {
+  if (reversed_) {
+    left_over = left_over->dagger();
+  }
+  EdgeVec preds;
+  op_signature_t sigs;
+  if (condition) {
+    left_over = std::make_shared<Conditional>(
+        left_over, (unsigned)condition->first.size(), condition->second);
+  }
+  Vertex new_v = circ_ptr_->add_vertex(left_over);
+  if (condition) {
+    for (const VertPort &vp : condition->first) {
+      preds.push_back(circ_ptr_->get_nth_out_edge(vp.first, vp.second));
+      sigs.push_back(EdgeType::Boolean);
+    }
+  }
+  preds.push_back(e);
+  sigs.push_back(EdgeType::Quantum);
+  circ_ptr_->rewire(new_v, preds, sigs);
+}
+
+bool SingleQubitSquash::is_squashable(Vertex v, OpType v_type) const {
+  return circ_ptr_->n_in_edges_of_type(v, EdgeType::Quantum) == 1 &&
+         is_gate_type(v_type) && squasher_->accepts(v_type);
+}
+
+bool SingleQubitSquash::sub_is_better(
+    const Circuit &sub, const std::vector<Gate_ptr> chain) const {
+  const unsigned n_gates = sub.n_gates();
+  return n_gates < chain.size() ||
+         (n_gates == chain.size() && !is_equal(sub, chain, reversed_));
```

Comment:
Can this ever lead to infinite sequences of squashes where the chain changes each time (cycling between values but staying the same length)?

Reply:
It all depends on the squasher. As long as you have a "normal form" which squashing leaves unchanged (which all well-behaved squasher should have) you'll be fine. In our case this is definitely the case. I can add a line in the docs saying that.